### ISSUE TITLE
.shop

### ISSUE THREAD

**Issue by AymenOthmani** [Mathew78540/realaion#22]
matthew do you have cmd .shop or need code it ? :)

**Comment by ghost** [comment]
Hello, I have created mine ; this is a first version with no special security or specific logging, but you can use it to test your shop.
Be carrefull, this has been created into 2.7 emulator of Lightning, you should probably adapt some functions and parameters.

Let me know if you see any error or if you have performance suggestion :)

```
package com.aionemu.gameserver.command.player;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.aionemu.commons.database.DB;
import com.aionemu.commons.database.DatabaseFactory;
import com.aionemu.commons.database.ParamReadStH;
import com.aionemu.gameserver.command.BaseCommand;
import com.aionemu.gameserver.dataholders.DataManager;
import com.aionemu.gameserver.model.gameobjects.player.Player;
import com.aionemu.gameserver.model.templates.item.ItemTemplate;
import com.aionemu.gameserver.services.SystemMailService;
import com.aionemu.gameserver.utils.PacketSendUtility;

/**
 * @author Sipherion
 */

public class CmdShop extends BaseCommand {

    public static Logger log = LoggerFactory.getLogger(CmdShop.class);

    public static final String DELETE_QUERY = "DELETE FROM `myshop` WHERE `id`=?";
    public static final String SELECT_QUERY = "SELECT `id`, `item`, `nb` FROM `myshop` WHERE `player_id`=?";

    public void execute(final Player player, String... params) {

        // There is no parameter needed
        if (params.length > 0) {
            showHelp(player);
            return;
        }

        final int playerId = player.getObjectId();
        final String playerName = player.getName();

        try {
            DB.select(SELECT_QUERY, new ParamReadStH() {

                // Set unique parameter to SQL query                
                public void setParams(PreparedStatement stmt) throws SQLException {
                    stmt.setInt(1, playerId);
                }

                // count will be used to inform player on how many objects he bought in the shop
                int count = 0;

                public void handleRead(ResultSet rset) throws SQLException {
                    // Loop on each result
                    while (rset.next()) {
                        count += 1;
                        int transactionId = rset.getInt("id");
                        int item = rset.getInt("item");
                        int nb = rset.getInt("nb");

                        // Catch item template for its name and maxStackCount
                        ItemTemplate itemTemplate = DataManager.ITEM_DATA.getItemTemplate(item);
                        int maxStackCount = (int) itemTemplate.getMaxStackCount();

                        // First of all, check if we can remove the row from database
                        if (removeShopTransaction(transactionId, playerId)) {
                            // These checks are used to see if there is more than maxStackCount to send
                            // If yes, multiple mail will be send to player
                            if (nb > maxStackCount && maxStackCount != 0) {
                                int qtyLeft = nb;
                                int qtySend = 0;
                                for(int i = 1; i <= nb ; i += maxStackCount) {
                                    if (qtyLeft > maxStackCount) {
                                        qtySend = maxStackCount;
                                        qtyLeft -= qtySend;
                                    }
                                    else {
                                        qtySend = qtyLeft;
                                    }
                                    SystemMailService.getInstance().sendMail(
                                            "WebShop", playerName, itemTemplate.getName(), "Voici votre commande (un seul courrier ne suffisait pas pour cette commande ; veuillez vérifier vos autres courriers).\nNous vous remercions pour votre achat.", 
                                            item, qtySend, 0, true
                                    );
                                }
                            }
                            else
                                SystemMailService.getInstance().sendMail("WebShop", playerName, itemTemplate.getName(), "Voici votre commande.\nNous vous remercions pour votre achat.", item, nb, 0, true);
                        }

                    }
                    // Inform player that there is no transaction waiting for this character
                    if(count == 0) {
                        PacketSendUtility.sendMessage(player, "Aucun achat boutique n'est en attente pour " + playerName);
                    }
                    else {
                        PacketSendUtility.sendMessage(player, "Les " + count + " achats effectues pour " + playerName + " ont bien ete recuperes.");
                    }
                }

            });
        }
        // Log error and inform player if there is a database error
        catch (Exception ex) {
            log.error("Could not catch transaction list for player " + playerId + " from DB: " + ex.getMessage(), ex);
            PacketSendUtility.sendMessage(player, "A database error has occured. Unable to satisfy your request.");
            return;
        }
    }

    // Function to remove transaction from database. Must be done first for security reasons.
    public boolean removeShopTransaction(int transactionId, int playerId) {
        Connection con = null;
        try {
            con = DatabaseFactory.getConnection();
            PreparedStatement stmt = con.prepareStatement(DELETE_QUERY);
            stmt.setInt(1, transactionId);
            stmt.execute();
            stmt.close();
        }
        catch (Exception e) {
            log.error("Could not delete transaction " + transactionId + " for player " + playerId + " from DB: " + e.getMessage(), e);
            return false;
        }
        finally {
            DatabaseFactory.close(con);
        }
        return true;
    }

}
```

**Comment by AymenOthmani** [comment]
tested it but give error 

`2016-07-31 08:24:24,324 Java Compiler ERROR: class CmdShop is public, should be declared in a file named CmdShop.java
Source: /C:/Users/Administrator/Desktop/AC-Game - pvp/./data/scripts/system/handlers/playercommands/cmd_shop.java
Line: 25
Column: 8
2016-07-31 08:24:24,762 Java Compiler ERROR: cannot find symbol
symbol  : class BaseCommand
location: package com.aionemu.gameserver.command
Source: /C:/Users/Administrator/Desktop/AC-Game - pvp/./data/scripts/system/handlers/playercommands/cmd_shop.java
Line: 14
Column: 38
2016-07-31 08:24:24,762 Java Compiler ERROR: cannot find symbol
symbol  : class SystemMailService
location: package com.aionemu.gameserver.services
Source: /C:/Users/Administrator/Desktop/AC-Game - pvp/./data/scripts/system/handlers/playercommands/cmd_shop.java
Line: 18
Column: 39
2016-07-31 08:24:24,762 Java Compiler ERROR: cannot find symbol
symbol: class BaseCommand
Source: /C:/Users/Administrator/Desktop/AC-Game - pvp/./data/scripts/system/handlers/playercommands/cmd_shop.java
Line: 25
Column: 30
2016-07-31 08:24:24,933 Java Compiler ERROR: cannot find symbol
symbol  : method showHelp(com.aionemu.gameserver.model.gameobjects.player.Player)
location: class com.aionemu.gameserver.command.player.CmdShop
Source: /C:/Users/Administrator/Desktop/AC-Game - pvp/./data/scripts/system/handlers/playercommands/cmd_shop.java
Line: 36
Column: 13
2016-07-31 08:24:24,933 Java Compiler ERROR: cannot find symbol
symbol: variable SystemMailService
Source: /C:/Users/Administrator/Desktop/AC-Game - pvp/./data/scripts/system/handlers/playercommands/cmd_shop.java
Line: 81
Column: 37
2016-07-31 08:24:24,933 Java Compiler ERROR: cannot find symbol
symbol: variable SystemMailService
Source: /C:/Users/Administrator/Desktop/AC-Game - pvp/./data/scripts/system/handlers/playercommands/cmd_shop.java
Line: 88
Column: 33

i will try to fix it, also im using AC-game 4.9.1 Version `

**Comment by mathieuletyrant** [comment]
``` java
package com.aionengine.gameserver.command.player;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


import com.aionemu.commons.database.DatabaseFactory;
import com.aionengine.gameserver.command.BaseCommand;
import com.aionengine.gameserver.model.gameobjects.player.Player;
import com.aionengine.gameserver.services.item.ItemService;
import com.aionengine.gameserver.utils.PacketSendUtility;

public class CmdShop extends BaseCommand {
    private static final Logger log = LoggerFactory.getLogger(CmdShop.class);

    public void execute(Player player, String... params) {
        Connection con = null;
        int nbitem = 0;
        try {
            con = DatabaseFactory.getConnection();
            PreparedStatement stmt = con.prepareStatement("SELECT * FROM myshop WHERE player_id = "+ player.getObjectId());
            PreparedStatement stmtDel;
            ResultSet resultSet = stmt.executeQuery();
            while (resultSet.next()) {
                int id = resultSet.getInt("id");
                int itemId = resultSet.getInt("item");
                int item_count = resultSet.getInt("nb");
                long count = ItemService.addItem(player, itemId, item_count);

                if (count == 0) {
                    stmtDel = con.prepareStatement("DELETE FROM myshop WHERE id = "+ id);
                    stmtDel.execute();
                    stmtDel.close();
                }
                else {
                    if (count != item_count) {
                        stmtDel = con.prepareStatement("UPDATE myshop SET nb = " + count + " WHERE id = " + id);
                        stmtDel.execute();
                        stmtDel.close();
                    }

                    PacketSendUtility.sendMessage(player, "Votre inventaire est plein, merci de faire de la place.");
                    return;
                }
                nbitem++;
            }
            resultSet.close();
            stmt.close();
        }
        catch (SQLException e) {
            log.error("command .shop", e);
        }
        finally {
            DatabaseFactory.close(con);
        }

        if (nbitem == 0)
            PacketSendUtility.sendMessage(player, "Vous n'avez acheter aucun Items en boutique.");
    }
}
```

**Comment by mathieuletyrant** [comment]
I have a lot a custom commands, but we have re-write commands system. So you must adapt this. But she works :)

**Comment by AymenOthmani** [comment]
already done created new cmd. and work :)
you can make your cmd work but if you have full emu, i using ac and they have web shop script need make some changes on it anyways thanks guys

**Comment by mathieuletyrant** [comment]
I close the issue, and i created a page on the Wiki : https://github.com/Mathew78540/realaion/wiki/Command-.shop
